### PR TITLE
Gdpr tasks

### PR DESCRIPTION
**Pull request by ximes**
- [x] Tiny amends (added colors to repo list, brewfile lock, missing repos)
- [x] GDPR deletion command 
- [x] Delete from Prime
- [x] Delete from Mixpanel
- [ ] Delete from Redshift
- [x] Delete from Audience

Goes with:
- https://github.com/mynewsdesk/mynewsdesk/pull/7183
- https://github.com/mynewsdesk/mnd-audience/pull/56

### CODE REVIEW

**Comment by dbackeus** [comment]
Please remove the Brewfile.lock thing. We should change the setup script to not generate it.

**Comment by ximes** [comment]
The system part is a bit annoying, but don't we want version locking on the taps and lang?

File: src/mnd/local_config.cr [hunk 1]
```diff
@@ -54,6 +54,66 @@ module Mnd
...
+    def mixpanel_user_project_api_project_token
+     @config["mixpanel_user_project_api_project_token"]
+    end
+
+    def mixpanel_user_project_api_project_token?
+     @config["mixpanel_user_project_api_project_token"]?
+    end
+
+    def mixpanel_user_project_api_project_token=(token)
+     @config["mixpanel_user_project_api_project_token"] = token
```
Comment (dbackeus) [inline src/mnd/local_config.cr:114]: Holy moly that's a lot of tokens...

Seems like you have some tabbing issue. Need one more space of indentation for all the new entries.

**Comment by dbackeus** [comment]
@ximes no, for brew we just assume that we always want the latest thing and that we're "likely backwards compatible" in general.

**Comment by dbackeus** [comment]
Hmm... after considering how much authorization stuff will be required to make this happen (have fun configuring the ssl tunnel + credentials for Redshift for example) maybe it would make more sense to create a separate web-service to do this instead that could hold all those tokens (also you could do that in Ruby which might be more comfy).

We could still have a command that runs a request to that API though.

But maybe the real value would be to have a GDPR deletion UI that support can handle without going through dev at all? 🤔

File: src/mnd/commands/gdpr/delete.cr [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+require "json"
+require "http/client"
+require "option_parser"
+
+module Mnd
+  class Commands::Support::Gdpr::Delete < Commands::Base
+    @dry_run : Bool = false
+
+    def initialize(args, dry_run = false)
+      @dry_run = dry_run
```
Comment (yxhuvud) [inline src/mnd/commands/gdpr/delete.cr:11]: It is a bit redundant to have dry_run take a default here, and also declare the variable above with a defualt. 

There is a bunch of alternatives here. I'd probably simply do 
```
class Foo < Whatevs
  def initialize(args, @dry_run = false)
    super(args)
  end
..
end
```

File: src/mnd/commands/gdpr/delete.cr [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+          email = confirmation[/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i]?
+          if email.nil?
+            display.warn "Invalid email. Any other?"
+            next
+          end
+
+          emails << "#{email}"
+        end
+      end
+      emails << "stefano.franzin@mynewsdesk.com"
```
Comment (yxhuvud) [inline src/mnd/commands/gdpr/delete.cr:36]: hardcoded to you?

File: src/mnd/commands/gdpr/delete.cr [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            auth_token: auth_token,
+            user_ids: distinct_ids,
+          )
+
+          begin
+            display.info "Task id for this removal is: #{results["results"]["task_id"]}"
+          rescue
+            display.warn "something went wrong: #{results.inspect}"
+          end
+
```
Comment (yxhuvud) [inline src/mnd/commands/gdpr/delete.cr:110]: You should run `crystal tool format`

File: src/mnd/local_config.cr [hunk 5]
```diff
@@ -54,6 +54,66 @@ module Mnd
...
+     @config["mixpanel_oauth_token"]
+    end
+
+    def mixpanel_oauth_token?
+     @config["mixpanel_oauth_token"]?
+    end
+
+    def mixpanel_oauth_token=(token)
+     @config["mixpanel_oauth_token"] = token
+    end
```
Comment (yxhuvud) [inline src/mnd/local_config.cr:67]: This is a bit repetetive. See https://crystal-lang.org/reference/syntax_and_semantics/macros.html , especially the section for iterating an ArrayLiteral. Perhaps a bit advanced though :)

**Comment by ximes** [comment]
Thanks for the first feedback. @yxhuvud it's a quick&dirty version, I didn't clean. I could optimize those getter/setter as well, etc. Thanks for the links, I'll check it out!
@dbackeus as I was finishing the mixpanel bit, I was thinking if we'd be better converting this to lambdas, or aws batches, exactly for that purpose, so support can just enqueue them themselves.
What's dragging me back is that putting those into lambdas would require us to map the db fields to remove, and to skip the rails logic we use in the deletion tasks. While EC2 would probably be overkill for this.
This was kind of a neutral middleground..... but it's ugly and noisy. Lots of work for a trivial task.
Another good option for me would be to add gdpr action in admin/advanced, which could trigger a local cleanse job + async tasks on other platforms. E.g. a anonymize lambda on aws, or a anonimyze endpoint on audience, etc. Just like this one is doing, but with this very step removed

**Comment by dbackeus** [comment]
@ximes why on earth would you use lambdas for this? Just create a rails app?

**Comment by dbackeus** [comment]
@ximes Using our current admin is an interesting idea. I guess it would provide us a good authorization mechanism out of the box. Then the actual deletion could be run directly on the production database partly, and perhaps setup an API app to handle the outlying services.

A dedicated GDPR rails app would have the advantage of likely making it easier to create a very custom UI experience though (like following progress of various jobs). But that might be overkill in any case...

**Comment by ximes** [comment]
To be precise: lambda as in a chain of: SQS > lambda processing stuff > deliver results, or similar. In general it's a good middle ground considering that we'll have to potentially connect to multiple dbs, process stuff, generate some file/report and deliver. The current-only gdpr task we have doesn't have that complexity, I was saying that we should use..... something external to prime to process.
Running api calls between rails apps will require us to do at least some planning and add authentication and controllers, hence lambdas. 
But a dedicated app is actually a good idea, however not sure if it's worth atm. I can leave the idea with Daniel - we may have other processes of which I'm not aware that could be worth wrapping in the same place?

**Comment by ximes** [comment]
btw, do we require ssl tunnelling of sorts for redshift? I think I can connect freely from dev?

**Comment by yxhuvud** [comment]
How does connecting to multiple databases change anything? You can do that just fine from heroku - you can even get integration so that multiple coupled apps that use a single db get their database config settings updated if the url changes. Which it sometimes can do as long as we are taking dbs residing in heroku.

**Comment by ximes** [comment]
Sure, but that wasn't just about db, but a list of things that can we done inside. We need some script capable or being enqueued, running, do three simple things and die. That can be, but doesn't have to be, a rack app and a server

**Comment by ximes** [comment]
PRIME-8849 is taking over putting everything in Prime. Closing.

**Pull request closed without merging**